doc/design.md

```diff
@@ -0,0 +1,175 @@
+
+* Any user deemed legitimate will be allowed to start a task
+* Only the user that started a task may interact with it, (e.g. query status)
+
+The obvious extension to this model is some form of Admin role that can query other users' tasks, but that is not being considered as part of this work.
+
+### System Integrity & Availability
+
+This system is assumed to be used by trusted users, and no effort will be made to prevent users damaging the system that is running these tasks. This includes actions like deleting resources, DoSing the system with a fork bomb, or any other harmful activity.
+
+Similarly, no effort will be made to prevent the user exfiltrating data from the server (e.g. running `cat ~/.ssh/id_rsa`).
+
+The above risks can be somewhat mitigated by restricting the permissions of the user running the API server.
+
+### Testing Considerations
+
+Both the server and client programs will need a configuration option to trust an arbitrary CA in order to run automated tests. Note that this specified CA *may* be hardcoded into the client and server binaries, depending on the time available for implementation. This is obviously something that would not be allowed in a production system.
+
+## Implementation Details
+
+### Server
+
+#### Process Execution
+Process execution will be managed via the Go `os.Process` type, most likely via the `os/exec` package. 
+
+No special effort will be made to ensure that all tasks are cleanly shut down if the server crashes. 
+
+#### Task Registration 
+There will be a single, central in-memory task registry, protected by a simple reader-writer lock. The assumption is that the values will be looked up more often created or destroyed, so the reader-writer should reduce contention on the registry.
+
+This has the value of simplicity, but it does mean that all task information is lost when the server is killed or crashes. It is easy to imagine this simple, in-memory mechanism being replaced by a database of some kind for increased durability.
+
+#### Individual Task Data 
+The tasks themselves will have their own, individual locking mechanism, so that concurrent requests on different tasks should not contend for the same lock (once the tasks are retrieved from the registry, that is).
+
+At present I am planning on using a reader/writer lock for each task, but this may change during implementation, depending on the complexity required and how much contention falls on each task.
+
+Logs from `stdout` and `stderr` will be treated as opaque binary data. The logs will also be stored in memory only. This again favours ease of implementation over durability, but it is easy to imagine a system where the logs are written to files or a database, and read back from there on request.
+
+#### User Database
+If required, any per-user data will be stored in a static config file loaded at startup. No modifications of any kind will be noticed during runtime.
```

Comment:
What kind of data will be stored in this config file?

Reply:
I don't actually know yet - because at this point I don't know enough about how gRPC authentication works to know exactly what info I will need.  For an exercise of this scope, "what I will need" may well be _nothing_, obviating the need for a user database or config file altogether.

I just wanted to be up-front that I will not be attempting to dynamically reload the server config during runtime.

I will re-word this paragraph to make the intent clearer.

Reply:
Done (hopefully)